### ISSUE TITLE
API Response Time [stream-input]

### ISSUE THREAD

**Issue by rodrigoGA** [elevenlabs/elevenlabs-python#134]
I am exploring the streaming functionality of Eleven Labs for enabling real-time conversations with a Large Language Model (LLM). I've come across information suggesting that the response times are typically under 1 second. However, during my tests, I've been experiencing response times exceeding 2 seconds. I am keen to understand how I might improve this.

- Behavior of 'optimize_streaming_latency' Variable: I noticed that the 'optimize_streaming_latency' variable is not considered when using the 'stream-input' api, but it seems to be functional in other scenarios. Is this a limitation of the library, or is it by design for websocket contexts?
-  Initial Websocket Packet Delay: When initializing the websocket, a packet containing the BOS variable and model details is transmitted. This process takes approximately 800ms. Is this duration typical, or are there ways to reduce it?
- Plan Impact on Response Times: I am currently on the "Creator" plan and in the prototype development phase. Could transitioning to a different plan offer better response times? If so, which plan would be most suitable for real-time conversations, considering my primary focus is on reducing latency rather than voice quality?
- Websocket Reusability for Multiple Streams: Given that my interactions with the LLM will be limited to 10-minute sessions, is it possible to keep the websocket open and send different streams through the same connection?
- Are there any suggested approaches to optimize response times, even if it requires compromising audio quality?

I appreciate your guidance on these matters. The website lacks detailed information on these specific queries, and I believe resolving these would greatly enhance the user experience for real-time applications.

**Comment by nicolapanegos** [comment]
Same issue, I usually have 4 second delay..

**Comment by bsridatta** [comment]
Use turbov2 model

**Comment by theanht1** [comment]
> Websocket Reusability for Multiple Streams

This is a valid point given the [7a. Limit the Number of Websocket Connection Closures](https://elevenlabs.io/docs/api-reference/reducing-latency#7a-limit-the-number-of-websocket-connection-closures) in the Reducing Latency document.

I think we can further improve the latency if we can reuse Eleven Labs websocket connection for multiple requests

**Comment by rodrigoGA** [comment]
> Use turbov2 model

Yes, the turbov2 model has much better response times, but as a downside, it is only available in English. However, all the previously listed issues still persist, with the major latency being in the sending of the BOS package.

**Comment by rodrigoGA** [comment]
> You can negate most if not all of the initial handshake latency by opening and initializing the WebSocket connection at the same time as you send the request to the LLM. Even if the LLM starts responding a bit earlier still, it's not until 50 characters (minimum) have been sent to ElevenLabs that audio generation can begin anyway, so it's fine if the handshake takes longer, as long as it's finished by the time you have received 50 characters from the LLM.

It could be a possible solution, but elevenlabs is opening the websocket and sending the BOS packet at the moment data starts being sent. That piece of code would have to be moved for the stream initialization. It's also important to consider that if such initialization is blocking, it would delay the start of the call to the llm, so the problem would only have been moved elsewhere.

**Comment by bephrem1** [comment]
@rodrigoGA Have you considered initializing the WebSocket connection before receiving user input? The first place my thinking goes is to get the initial TLS handshake out of the way off the path of usage (maybe with dummy messages, etc) — just to ensure the connection is ready to go w/ no overhead at request time.

**Comment by drevicko** [comment]
> > You can negate most if not all of the initial handshake latency by opening and initializing the WebSocket connection at the same time as you send the request to the LLM. Even if the LLM starts responding a bit earlier still, it's not until 50 characters (minimum) have been sent to ElevenLabs that audio generation can begin anyway, so it's fine if the handshake takes longer, as long as it's finished by the time you have received 50 characters from the LLM.
> 
> It could be a possible solution, but elevenlabs is opening the websocket and sending the BOS packet at the moment data starts being sent. That piece of code would have to be moved for the stream initialization. It's also important to consider that if such initialization is blocking, it would delay the start of the call to the llm, so the problem would only have been moved elsewhere.

This is not the case. If you look at the SDK code under `generate_stream_input()` in `tts.py` (v 2.27 - likely changed in v3.0b) you'll see that the websocket is immediately created and BOS sent before reading from the provided text iterator.

**Comment by RohitMidha23** [comment]
@rodrigoGA can you share some code snippets of how you are handling this case? Seem to be running into a lot more latency here and would love to identify what's going wrong!

**Comment by rodrigoGA** [comment]
I have moved this question to Discord and have obtained some more information from https://discord.com/channels/1066739690436313158/1208052700441219072/1208052700441219072 which continues in an interesting thread by @bephrem1  https://discord.com/channels/1066739690436313158/1212116224385884242.

In my opinion, the WebSocket API can be flexible even when used from a browser, but it seems to add more problems than solutions in backend real-time applications (which I understand it is intended for) in terms of connection reuse, latencies, etc. It would be great if 11Labs could provide a grpc API for applications that require streaming of data with very low latency.

**Comment by AngeloGiacco** [comment]
hi all! thanks for the great responses here. i'm closing this as it is not directly relevant to our SDK, and will note that now the preferred model is https://elevenlabs.io/docs/models#flash-v25
